packages/next/src/server/revalidation-utils.ts

```diff
@@ -63,16 +67,55 @@ function diffRevalidationState(
   }
 }
 
-async function executeRevalidates(
+async function revalidateTags(
+  tags: string[],
+  incrementalCache: IncrementalCache | undefined
+): Promise<void> {
+  if (tags.length === 0) {
+    return
+  }
+
+  const promises: Promise<void>[] = []
+
+  if (incrementalCache) {
+    promises.push(incrementalCache.revalidateTag(tags))
+  }
+
+  const handlers = getCacheHandlers()
+  if (handlers) {
+    for (const handler of handlers) {
+      promises.push(handler.expireTags(...tags))
+    }
+  }
+
+  await Promise.all(promises)
+
+  const workUnitStore = workUnitAsyncStorage.getStore()
+
+  if (workUnitStore?.implicitTags) {
+    const tagsSet = new Set(tags)
+
+    if (workUnitStore.implicitTags.tags.some((tag) => tagsSet.has(tag))) {
+      await updateImplicitTagsExpiration(workUnitStore.implicitTags)
+    }
+  }
+}
+
+export async function executeRevalidates(
   workStore: WorkStore,
-  {
-    pendingRevalidatedTags,
-    pendingRevalidates,
-    pendingRevalidateWrites,
-  }: RevalidationState
+  state?: RevalidationState
```

Comment:
Seems like this uses the `workStore` or the `state`, why not just accept the `state` as the only parameter so that this `??` is done at the callsite?

Reply:
I had that initially, but didn't want to repeat it for four different call sites.

Reply:
And then there's also the fact, that we need to pass the work store anyways (or at least its `incrementalCache` property).